url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/power/PowerVm.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Title:        CloudSim Toolkit
+ * Description:  CloudSim (Cloud Simulation) Toolkit for Modeling and Simulation of Clouds
+ * Licence:      GPL - http://www.gnu.org/copyleft/gpl.html
+ *
+ * Copyright (c) 2009-2012, The University of Melbourne, Australia
+ */
+
+package org.cloudbus.cloudsim.power;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import org.cloudbus.cloudsim.CloudletScheduler;
+import org.cloudbus.cloudsim.Vm;
+import org.cloudbus.cloudsim.core.CloudSim;
+import org.cloudbus.cloudsim.util.MathUtil;
+
+import flanagan.analysis.Stat;
+
+/**
+ * The class of a VM that stores its CPU utilization history. The history is used by VM allocation
+ * and selection policies.
+ * 
+ * If you are using any algorithms, policies or workload included in the power package please cite
+ * the following paper:
+ * 
+ * Anton Beloglazov, and Rajkumar Buyya, "Optimal Online Deterministic Algorithms and Adaptive
+ * Heuristics for Energy and Performance Efficient Dynamic Consolidation of Virtual Machines in
+ * Cloud Data Centers", Concurrency and Computation: Practice and Experience, ISSN: 1532-0626, Wiley
+ * Press, New York, USA, 2011, DOI: 10.1002/cpe.1867
+ * 
+ * @author Anton Beloglazov
+ * @since CloudSim Toolkit 2.0
+ */
+public class PowerVm extends Vm {
+
+	/** The Constant HISTORY_LENGTH. */
+	public static final int HISTORY_LENGTH = 30;
+
+	/** The utilization history. */
+	private final List<Double> utilizationHistory = new LinkedList<Double>();
+
+	/** The previous time. */
+	private double previousTime;
+
+	/** The scheduling interval. */
+	private double schedulingInterval;
+
+	/**
+	 * Instantiates a new power vm.
+	 * 
+	 * @param id the id
+	 * @param userId the user id
+	 * @param mips the mips
+	 * @param pesNumber the pes number
+	 * @param ram the ram
+	 * @param bw the bw
+	 * @param size the size
+	 * @param priority the priority
+	 * @param vmm the vmm
+	 * @param cloudletScheduler the cloudlet scheduler
+	 * @param schedulingInterval the scheduling interval
+	 */
+	public PowerVm(
+			int id,
+			int userId,
+			double mips,
+			int pesNumber,
+			int ram,
+			long bw,
+			long size,
+			int priority,
+			String vmm,
+			CloudletScheduler cloudletScheduler,
+			double schedulingInterval) {
+		super(id, userId, mips, pesNumber, ram, bw, size, vmm, cloudletScheduler);
+		setSchedulingInterval(schedulingInterval);
+	}
+
+	/**
+	 * Updates the processing of cloudlets running on this VM.
+	 * 
+	 * @param currentTime current simulation time
+	 * @param mipsShare array with MIPS share of each Pe available to the scheduler
+	 * 
+	 * @return time predicted completion time of the earliest finishing cloudlet, or 0 if there is
+	 *         no next events
+	 * 
+	 * @pre currentTime >= 0
+	 * @post $none
+	 */
+	@Override
+	public double updateVmProcessing(double currentTime, List<Double> mipsShare) {
+		double time = super.updateVmProcessing(currentTime, mipsShare);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;mipsShare&#39; (index: 1 | type: List). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;currentTime&#39;. The same parameter type is validated here for example: url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/Examples/Dvfs_example_more_complex.java:202. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).